File: guessthedatastructure.java
import java.io.*;
import java.util.*;

public class guessthedatastructure {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        String str = " ";
        while ((str = br.readLine()) != null){// check for EOF
            int n = Integer.parseInt(str);
            Stack<Integer> S = new Stack<>();
            Queue<Integer> Q = new LinkedList<>();
            PriorityQueue<Integer> PQ = new PriorityQueue<>(Collections.reverseOrder());
            HashSet<String> dataStructures = new HashSet<>();
            dataStructures.add("Stack"); dataStructures.add("Queue"); dataStructures.add("Priority Queue");
            boolean impossible = false;
            for (int i = 0; i < n; i++){
                String[] line = br.readLine().split(" ");
                int command = Integer.parseInt(line[0]), element = Integer.parseInt(line[1]);
                if (command == 1){
                    S.push(element);
                    Q.offer(element);
                    PQ.add(element);
                }
                else {
                    if (S.size() == 0 || Q.size() == 0 || PQ.size() == 0){impossible = true;}
                    else {
                        int a = S.pop(), b = Q.poll(), c = PQ.remove();
                        if (a != element) {dataStructures.remove("Stack");}
                        if (b != element) {dataStructures.remove("Queue");}
                        if (c != element) {dataStructures.remove("Priority Queue");}
                    }
                }
            }
            if (impossible || dataStructures.size() == 0) {pw.println("impossible");}
            else if (dataStructures.size() == 1) {pw.println(dataStructures.toArray()[0]);}
            else if (dataStructures.size() > 1) {pw.println("not sure");}
        }
        br.close();
        pw.close();
    }
}

File: guessthedatastructure.py
from sys import stdin,stdout
from collections import deque
from heapq import heappop, heappush

while True:
    try:
        n = int(stdin.readline())
    except:
        break
    stack,queue,PQ,impossible = [],deque([]),[],False
    data_structures = {'stack','queue','priority queue'}
    for _ in range(n):
        command,element = map(int,stdin.readline().split())
        if command == 1:
            stack.append(element)
            queue.append(element)
            heappush(PQ,-element) #since PQ in python is min heap by default
        else:
            if not stack or not queue or not PQ: # check if empty
                impossible = True
            else:
                a,b,c = stack.pop(),queue.popleft(),-heappop(PQ)
                if a != element:
                    data_structures.discard('stack')
                if b != element:
                    data_structures.discard('queue')
                if c != element:
                    data_structures.discard('priority queue')
    if impossible or len(data_structures) == 0:
        stdout.write("impossible\n")
    elif len(data_structures) == 1:
        stdout.write(f'{list(data_structures)[0]}\n')
    elif len(data_structures) > 1:
        stdout.write(f'not sure\n')
    
